### PR TITLE
 Support for out of order processing of replication tasks

### LINKED ISSUE #672: Support for out of order processing of replication events

### PR DESCRIPTION
**Pull request by samarabbas**
Schema changes to buffer replication tasks as part of mutable state.
Persistence layer changes to provide support for adding new replication
tasks, and deleting replication tasks after they are processed and
applied to workflow execution.

MutableState change to support CRUD of buffered replication tasks.  And
HistoryReplicator changes to detect out-of-order events and instead add
replication task to the buffer.  Also loop through all replication task
and flush any buffered tasks after processing of an update.

fixes #672

### CODE REVIEW

File: common/persistence/cassandraPersistence.go [hunk 1]
```diff
@@ -2804,6 +2928,28 @@ func createSignalInfo(result map[string]interface{}) *SignalInfo {
...
+		case "first_event_id":
+			info.FirstEventID = v.(int64)
+		case "next_event_id":
+			info.NextEventID = v.(int64)
+		case "version":
+			info.Version = v.(int64)
+		case "history":
+			h := v.(map[string]interface{})
+			info.History = createSerializedHistoryEventBatch(h)
+		case "new_run_history":
```
Comment (wxing1292) [inline common/persistence/cassandraPersistence.go]: what is this new run history?

File: common/persistence/cassandraPersistence_test.go [hunk 2]
```diff
@@ -1754,3 +1943,21 @@ func copyReplicationInfo(sourceInfo *ReplicationInfo) *ReplicationInfo {
 		LastEventID: sourceInfo.LastEventID,
 	}
 }
+
+func serializeHistoryEvents(events []*gen.HistoryEvent) *SerializedHistoryEventBatch {
```
Comment (wxing1292) [inline common/persistence/cassandraPersistence_test.go]: can you add (s *cassandraPersistenceSuite)  so this function will not be used by src code?

File: common/persistence/cassandraPersistence_test.go [hunk 3]
```diff
@@ -1754,3 +1943,21 @@ func copyReplicationInfo(sourceInfo *ReplicationInfo) *ReplicationInfo {
...
+	historySerializer := NewJSONHistorySerializer()
+	bufferedBatch := NewHistoryEventBatch(GetDefaultHistoryVersion(), events)
+	serializedEvents, _ := historySerializer.Serialize(bufferedBatch)
+
+	return serializedEvents
+}
+
+func deserializedHistoryEvents(batch *SerializedHistoryEventBatch) *HistoryEventBatch {
```
Comment (wxing1292) [inline common/persistence/cassandraPersistence_test.go]: can you add (s *cassandraPersistenceSuite)  so this function will not be used by src code?

File: common/persistence/dataInterfaces.go [hunk 4]
```diff
@@ -511,20 +521,22 @@ type (
...
+		DeleteChildExecutionInfo      *int64
+		UpsertRequestCancelInfos      []*RequestCancelInfo
+		DeleteRequestCancelInfo       *int64
+		UpsertSignalInfos             []*SignalInfo
+		DeleteSignalInfo              *int64
+		UpsertSignalRequestedIDs      []string
+		DeleteSignalRequestedID       string
+		NewBufferedEvents             *SerializedHistoryEventBatch
+		ClearBufferedEvents           bool
+		NewBufferedReplicationTask    *BufferedReplicationTask
```
Comment (wxing1292) [inline common/persistence/dataInterfaces.go]: should we use a slice? or is NewBufferedReplicationTask and DeleteBufferedReplicationTask guaranteed to have at most one at a time?

File: service/history/historyReplicator.go [hunk 5]
```diff
@@ -88,7 +89,42 @@ func (r *historyReplicator) ApplyEvents(request *h.ReplicateEventsRequest) (retE
...
+			if t := msBuilder.BufferReplicationTask(request); t == nil {
+				return errors.New("failed to add buffered replication task")
+			}
+
+			return nil
+		}
+	}
+
+	// First check if there are events which needs to be flushed before applying the update
+	err = r.FlushBuffer(context, msBuilder, request)
```
Comment (wxing1292) [inline service/history/historyReplicator.go]: the flush buffer function will internally call ApplyReplicationTask, so is there any redundant logic here?

File: service/history/mutableStateBuilder.go [hunk 6]
```diff
@@ -291,10 +300,86 @@ func (e *mutableStateBuilder) CloseUpdateSession() (*mutableStateSessionUpdates,
...
 	}
+	e.updateBufferedReplicationTasks = nil
+	e.deleteBufferedReplicationEvent = nil
 
 	return updates, nil
 }
 
+func (e *mutableStateBuilder) BufferReplicationTask(
+	request *h.ReplicateEventsRequest) *persistence.BufferedReplicationTask {
+	if _, ok := e.GetBufferedReplicationTask(request.GetFirstEventId()); ok {
```
Comment (wxing1292) [inline service/history/mutableStateBuilder.go]: what if there are 2 active cluster?
2 can sent out a batch of events which have the same first event ID, i believe a comparison of failover version is needed?

**Comment by wxing1292** [comment]
if travis could not build, just change this line:
https://github.com/uber/cadence/blob/master/.travis.yml#L26

File: common/persistence/cassandraPersistence.go (hunk 1 above)
Comment (samarabbas) [reply common/persistence/cassandraPersistence.go]: Replication Task for the ContinueAsNew has the history for both current run and new run as the same transaction.

File: common/persistence/cassandraPersistence_test.go (hunk 2 above)
Comment (samarabbas) [reply common/persistence/cassandraPersistence_test.go]: Done.

File: common/persistence/cassandraPersistence_test.go (hunk 3 above)
Comment (samarabbas) [reply common/persistence/cassandraPersistence_test.go]: Done.

File: common/persistence/dataInterfaces.go (hunk 4 above)
Comment (samarabbas) [reply common/persistence/dataInterfaces.go]: Currently we are applying replication task as a single transaction, so there is no need currently.  We will cross that bridge when there is a need for batching them together.

File: service/history/historyReplicator.go (hunk 5 above)
Comment (samarabbas) [reply service/history/historyReplicator.go]: Flush calls it for buffered replication tasks, not for the current one being processed.

File: service/history/mutableStateBuilder.go (hunk 6 above)
Comment (samarabbas) [reply service/history/mutableStateBuilder.go]: Yes.  It will be handled once I start implementing the merge protocol.

**Pull request merged**